server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -269,6 +278,50 @@ public void parse(ParseContext context) throws IOException {
      */
     protected abstract void parseCreateField(ParseContext context) throws IOException;
 
+    /**
+     * Given access to a document's _source, return this field's values.
+     *
+     * In addition to pulling out the values, mappers can parse them into a standard form. This
+     * method delegates parsing to {@link #parseSourceValue} for parsing. Most mappers will choose
+     * to override {@link #parseSourceValue} -- for example numeric field mappers make sure to
+     * parse the  source value into a number of the right type.
+     *
+     * Some mappers may need more flexibility and can override this entire method instead.
+     *
+     * @param lookup a lookup structure over the document's source.
+     * @param format an optional format string used when formatting values, for example a date format.
+     * @return a list a standardized field values.
+     */
+    public List<?> lookupValues(SourceLookup lookup, @Nullable String format) {
```

Comment:
I wonder if the returned list should be sorted similarly to what we do for `doc_values` or if we should preserve the original order in the document's source ? I am saying this since we may want to return results from doc_values in the future but that would be inconsistent with the current way of retrieving fields.

Reply:
I had thought that the order of the returned list was "undefined". They come back in the sort order as the `_source` now, but that is more an accident of the implementation than anything.

Reply:
In my experience, users don't find that doc values behavior to be useful and are sometimes quite confused by it. One example is https://github.com/elastic/elasticsearch/issues/29110, and I've also seen discuss issues asking why doc values don't preserve the order of the array. I wonder if sorting the results by value will it harder for users to switch to the API from source filtering, in cases where they care about returning the original list order.

This API can also return structured values like numeric ranges or geo points, which don't have an obvious default sort order. It could be harder to understand the behavior if arrays sometimes came back in sorted order, sometimes not?

Reply:
I think the safest thing may be to document that the order of the returned results is undefined. Maybe in a follow up we could add a parameter to preserve the order of the original document? If someone asks for it. That'd just force us to `_source`.

Reply:
> harder for users to switch to the API from source filtering

Agreed, I was more seeing this from users that switch to the API from `docvalues_field` but I can see how it could be confusing coming from source filtering. We don't have clear plans to retrieve values from doc_values in this API so I don't have strong feeling here. Let's keep it undefined for now and we can discuss again if we add something that uses doc values under the hood ?

Reply:
+1 to what @nik9000 said, his comment appeared after I posted mine so I am glad we came to the same conclusion ;).

Reply:
>I think the safest thing may be to document that the order of the returned results is undefined. 
> Let's keep it undefined for now and we can discuss again if we add something that uses doc values under the hood ?

This works for me, I'll add a note to the docs that users shouldn't rely on the order of the values.

Reply:
Oops, multiple race conditions here :) Sounds like we are all on the same page.

Reply:
Three way race!